libursa/QueryGraph.cpp

```diff
@@ -23,6 +31,84 @@ QueryGraph QueryGraph::dual() const {
     return result;
 }
 
+class NodeState {
+    std::vector<NodeId> ready_predecessors_;
+    uint32_t total_predecessors_;
+    QueryResult state_;
+
+   public:
+    NodeState() : state_(QueryResult::everything()), total_predecessors_(0) {}
+
+    const QueryResult &state() const { return state_; }
+
+    const std::vector<NodeId> &ready_predecessors() const {
+        return ready_predecessors_;
+    }
+
+    void set(QueryResult &&state) { state_ = std::move(state); }
+
+    void add_predecessor() { total_predecessors_++; }
+
+    void add_ready_predecessor(NodeId ready) {
+        ready_predecessors_.push_back(ready);
+    }
+
+    bool ready() const {
+        return ready_predecessors_.size() >= total_predecessors_;
+    }
+};
+
+QueryResult masked_or(std::vector<const QueryResult *> *to_or,
+                      QueryResult &&mask) {
+    if (to_or->empty()) {
+        return std::move(mask);
+    }
+    QueryResult result{QueryResult::empty()};
+    for (auto query : *to_or) {
+        // TODO(msm): we should do everything in parallel here.
+        QueryResult alternative{query->vector()};
+        alternative.do_and(mask);
+        result.do_or(std::move(alternative));
+    }
+    return result;
+}
+
+QueryResult QueryGraph::run(const QueryFunc &oracle) const {
```

Comment:
Is it not a strange fate that we should suffer so much fear and doubt for so small a thing?